src/vmm/src/vstate/vm/mod.rs

```diff
@@ -55,6 +63,26 @@ impl Vm {
         Ok(Vm { fd, arch })
     }
 
+    /// Creates the specified number of [`Vcpu`]s.
+    ///
+    /// The returned [`EventFd`] is written to whenever any of the vcpus exit.
+    pub fn create_vcpus(&mut self, vcpu_count: u8) -> Result<(Vec<Vcpu>, EventFd), VmError> {
+        self.arch_pre_create_vcpus(vcpu_count)?;
+
+        let exit_evt = EventFd::new(libc::EFD_NONBLOCK).map_err(VmError::EventFd)?;
+
+        let mut vcpus = Vec::with_capacity(vcpu_count as usize);
+        for cpu_idx in 0..vcpu_count {
+            let exit_evt = exit_evt.try_clone().map_err(VmError::EventFd)?;
+            let vcpu = Vcpu::new(cpu_idx, self, exit_evt).map_err(VmError::CreateVcpu)?;
+            vcpus.push(vcpu);
+        }
+
+        self.arch_post_create_vcpus(vcpu_count)?;
+
+        Ok((vcpus, exit_evt))
+    }
```

Comment:
Maybe just copy it for each arch. This one loop is not worth doing this `pre_*`/`post_*` shenanigans.

Reply:
This one I prefer the pre_*/post_* thingy over duplicating the function. It's how the kernel implements these things, and it seems to work pretty well over there